rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,171 @@
+process being created for each SSH connection.
+
+This has a few drawbacks:
+
+- The `tbot` binary has a significant memory overhead, much of which is not
+  necessary for proxying the SSH connection. When connecting to a large number
+  of hosts, this overhead can become significant.
+- A process per connection, with each process having a number of tasks, places
+  the Linux scheduler under significant load.
+- Each `tbot` process must perform a number of initial startup tasks
+  individually, contributing to latency in establishing the SSH tunnel. Some of 
+  these tasks could be shared, or their results cached.
+- Monitoring a large number of short-lived `tbot` processes is difficult.
+- There is no easy way to implement global ratelimits and backoffs across all
+  the short-lived `tbot` processes. This makes gracefully handling upstream
+  outages problematic.
+
+## Details
+
+### Implementation
+
+We will introduce a long-lived service in `tbot`: `ssh-local-proxy`.
+
+Upon start, the service will write a SSH config to a configured destination.
+This SSH config should be used by OpenSSH to leverage the service.
+
+Once initialized, the service will listen on a Unix socket. Upon connection to this
+socket, the client will write a short message containing the target host and
+port. The service will then establish the connection, and then begin forwarding
+data between the socket and the upstream connection.
+
+The client will be configured as the `ProxyCommand`, and additionally
+`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
+to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
+actively forwarding the connection. This means the client can exit after the
+connection to the local proxy is established.
+
+This solution means that whilst a process will be spun up for each SSH
+connection, it will return after the connection is established. This reduces the
+long-term overhead associated with each connection and the impact on the Linux
+scheduler.
+
+#### Protocol
+
+The Unix socket will implement a simple protocol:
+
+1. Upon connection, the client will send a message indicating the target.
+  a. The message will be encoded in JSON. This will allow more fields to be 
+    added in the future without breaking compatability.
+  b. The `host_port` field will be a string in the format `host:port`.
+2. The server will establish a connection to the target, and then begin
+   forwarding data between the local connection and the target.
+
+Also considered was implementing the SOCKS5 protocol. Whilst this is more 
+standardized, it is more complicated than necessary. It also relies on our
+existing implementation of a SOCKS5 server which has not received signficant
+attention in some time.
+
+#### Client Binary
+
+As it stands today, the `tbot` binary has significant overhead associated with
+it.
+
+Due to this, we will introduce a new, extremely lightweight binary that will
+act as the client.
```

Comment:
I discussed this with Edoardo - since there's no reason we could have a subcommand by default and then users can switch to the more performant standalone if they want. It'll definitely put a dink in the performance but I'm unsure if @espadolini has done a measurement on that.